docs/articles/improve-raptor-with-rag.md

```diff
@@ -0,0 +1,561 @@
+    results = {}
+    df_clusters, df_summaries = embed_cluster_summaries(texts, level, target_dim, threshold)
+    results[level] = (df_clusters, df_summaries)
+    
+    if df_summaries.empty or len(df_summaries['clusters'].unique()) == 1:
+        print(f"No more unique clusters found at level {level}. Stopping recursion.")
+        return results
+    
+    if level < number_of_levels:
+        next_level_texts = df_summaries['summaries'].tolist()
+        next_level_results = recursive_embedding_with_cluster_summarization(
+            next_level_texts, 
+            number_of_levels, 
+            level + 1,
+            target_dim,
+            threshold
+        )
+        results.update(next_level_results)
+    
+    return results
+ 
+```
+
+Okay, the code might seem a bit daunting at first glance, but don’t worry! Just give it a couple of looks, and it will start to make sense. Essentially, we’re just following the flow I mentioned earlier. 
+
+```python
+def process_text_hierarchy(
+    texts: List[str], 
+    number_of_levels: int = 3,
+    target_dim: int = 10,
+    threshold: float = 0.1
+) -> Dict[str, pd.DataFrame]:
+    hierarchy_results = recursive_embedding_with_cluster_summarization(
+        texts, number_of_levels, target_dim=target_dim, threshold=threshold
+    )
+    
+    processed_results = {}
+    for level, (df_clusters, df_summaries) in hierarchy_results.items():
+        if df_clusters.empty or df_summaries.empty:
+            print(f"No data for level {level}. Skipping.")
+            continue
+        processed_results[f"level_{level}_clusters"] = df_clusters
+        processed_results[f"level_{level}_summaries"] = df_summaries
+    
+    return processed_results
+
+results = process_text_hierarchy(chunks, number_of_levels=3)
+```
+
+![image.png](../assets/use_cases/improve-raptor-with-rag/raptor-3.png)
+
+### Inference
+
+Now that we have our tree structure with leaf nodes at the bottom and summarized nodes in between, it’s time to query the RAG. There are two main methods for navigating the RAPTOR tree: Tree Traversal and Collapsed Tree Retrieval.
+
+1. **Tree Traversal Retrieval:** This method systematically explores the tree, starting from the root node. It first selects the top-k most relevant root nodes based on their cosine similarity to the query embedding. Then, for each selected root node, its children are considered in the next layer, where the top-k nodes are again selected based on their cosine similarity to the query vector. This process repeats until we reach the leaf nodes. Finally, the text from all the selected nodes is concatenated to form the retrieved context.
+2. **Collapsed Tree Retrieval:** This approach simplifies things by viewing the tree as a single layer. Here, it directly compares the query embedding to the vector embeddings of all the leaf nodes (the original text chunks) and summary nodes. This method works best for factual, keyword-based queries where you need specific details.
+
```

Comment:
I thought the point of using RAPTOR was NOT to flatten but instead preserve the hierarchy... I would have thought a use case that requires tree traversal rather than collapsed tree would make more sense here. Can you explain? 
We need to provide a better rationale for
1 choosing a query that uses collapsed tree
2 why, in a little more detail, collapsed tree is better in this case than tree traversal
?

Reply:
There are three major points that I can think of for choosing the collapsed tree over the hierarchy one for this time. 

1. Nature of the query: So basically the Collapsed trees tend to work well when we need a quick response and when the hierarchy is considered, it introduces a bit of complexity. In cases where the hierarchy isn't essential to understanding the answer (i.e., the query doesn't demand complex multi-level reasoning), collapsing the tree makes retrieval faster and more efficient. 

2. Speeding the things:  Well as I worked on the collapsed tree for this time, my main aim was to avoid the overhead of navigating each hierarchical level. This significantly improves the retrieval speed, especially in large document collections where traversing multiple levels could introduce a bit of latency. 

3. Efficiency in Retrieval : Well the most important use case I can think when implementing the collapsed one for now is due to the reason that summarized nodes from higher levels are treated as additional chunks for us, it means that we can pull in higher-level summaries alongside granular details in just one pass. This works good when the query needs the context from the granular as well as a global perspective considering the various summarized nodes. So for the financial docs, it's necessary to get the overall perspective too regarding the deals, quarter results along with the pin points node information 

Let me know Robert if this was helpful